src/fetch-map/source.ts

```diff
@@ -0,0 +1,223 @@
+  }
+
+  const spatialIndexOptions = {
+    aggregationExp: !aggregationExp ? DEFAULT_AGGREGATION_EXP : aggregationExp,
+    aggregationResLevel: scaleAggregationResLevel(
+      aggregationResLevel,
+      tileResolution
+    ),
+    spatialDataColumn,
+    ...(filters && {filters}),
+  } as H3QuerySourceOptions;
+  const tilesetOptions = {
+    ...sourceOptions,
+    tableName: source,
+  } as VectorTilesetSourceOptions;
+  const tableOptions = {
+    ...sourceOptions,
+    tableName: source,
+    tileResolution,
+  } as TableSourceOptions;
+  const queryOptions = {
+    ...sourceOptions,
+    sqlQuery: source,
+    tileResolution,
+    ...(queryParameters && {queryParameters}),
+  } as QuerySourceOptions;
+  const vectorOptions = {
+    spatialDataColumn,
+    ...(columns && {columns}),
+    ...(filters && {filters}),
+    ...(aggregationExp && {aggregationExp}),
+  } as VectorTableSourceOptions;
+
+  if (type === 'raster') {
+    return rasterSource({
+      ...sourceOptions,
+      tableName: source,
+      ...(filters && {filters: filters as any}),
+    });
+  }
+  if (type === 'tileset') {
+    return vectorTilesetSource({...tilesetOptions});
+  }
+
+  if (type === 'table') {
+    if (isH3) {
+      return h3TableSource({
+        ...(tableOptions as H3TableSourceOptions),
+        ...spatialIndexOptions,
+      });
+    } else if (isQuadbin) {
+      return quadbinTableSource({
+        ...(tableOptions as QuadbinTableSourceOptions),
+        ...spatialIndexOptions,
+      });
+    } else {
+      return vectorTableSource({
+        ...(tableOptions as VectorTableSourceOptions),
+        ...vectorOptions,
+      });
+    }
+  }
+
+  if (type === 'query') {
+    if (isH3) {
+      return h3QuerySource({
+        ...(queryOptions as H3QuerySourceOptions),
+        ...spatialIndexOptions,
+      });
+    } else if (isQuadbin) {
+      return quadbinQuerySource({
+        ...(queryOptions as QuadbinQuerySourceOptions),
+        ...spatialIndexOptions,
+      });
+    } else {
+      return vectorQuerySource({
+        ...(queryOptions as VectorQuerySourceOptions),
+        ...vectorOptions,
+      });
+    }
+  }
+  throw new Error(`Invalid source type: ${type}`);
+}
+
+function getSourceOptions({
+  accessToken,
+  apiBaseUrl,
+  connection,
+  headers,
+  maxLengthURL,
+}: FetchDatasetOptions) {
+  return {
+    accessToken,
+    connectionName: connection,
+    apiBaseUrl,
+    headers,
+    maxLengthURL,
+    ...(headers?.['Cache-Control']?.includes('no-cache') && {
+      localCache: {
+        cacheControl: ['no-cache'] as 'no-cache'[],
+      },
+    }),
+  };
+}
+
+function getDynamicTileResolution(
+  spatialIndex?: SpatialIndex | null
+): TileResolution {
+  // TODO: Support increased tile size and resolution for dynamic H3 spatial indexes.
+  if (spatialIndex !== SpatialIndex.H3) {
+    return REDUCED_QUERIES_TILE_RESOLUTION;
+  }
+
+  return DEFAULT_TILE_RESOLUTION;
+}
+
+function scaleAggregationResLevel(
```

Comment:
Could you include the comments from the source files? I worry some of this code will be difficult to parse for future us.

```suggestion
/**
 * State of `aggregationResLevel` in the UI and backend config is based on an assumption of
 * 512x512px tiles. Because we may change tile resolution for performance goals, the
 * `aggregationResLevel` passed to the deck.gl layer must be scaled with tile resolution.
 */
function scaleAggregationResLevel(
```